drivers/SmartThings/zwave-garage-door-opener/src/ecolink-zw-gdo/init.lua

```diff
@@ -0,0 +1,458 @@
+--- @type st.zwave.CommandClass.Notification
+local Notification = (require "st.zwave.CommandClass.Notification")({ version = 8 })
+
+-- Ecolink garage door operator
+local ECOLINK_GARAGE_DOOR_FINGERPRINTS = {
+  { manufacturerId = 0x014A, productType = 0x0007, productId = 0x4731 }
+}
+
+local GDO_ENDPOINT_NAME = "main"
+local CONTACTSENSOR_ENDPOINT_NAME = "sensor"
+local GDO_ENDPOINT_NUMBER = 1
+local CONTACTSENSOR_ENDPOINT_NUMBER = 2
+
+local CONTACTSENSOR_BATTERY_LEVEL_NORMAL = 100
+local CONTACTSENSOR_BATTERY_LEVEL_LOW = 1
+local CONTACTSENSOR_BATTERY_LEVEL_DEAD = 0
+
+local GDO_CONFIG_PARAM_NO_UNATTENDED_WAIT = 1
+local GDO_CONFIG_PARAM_NO_ACTIVATION_TIME = 2
+local GDO_CONFIG_PARAM_NO_OPEN_TIMEOUT = 3
+local GDO_CONFIG_PARAM_NO_CLOSE_TIMEOUT = 4
+local GDO_CONFIG_PARAM_NO_SHAKE_SENSE = 5
+local GDO_CONFIG_PARAM_NO_APP_RETRY = 6
+
+--- Determine whether the passed device is an Ecolink garage door operator
+---
+--- @param driver Driver driver instance
+--- @param device Device device isntance
+--- @return boolean true if the device proper, else false
+local function can_handle_ecolink_garage_door(opts, driver, device, ...)
+  for _, fingerprint in ipairs(ECOLINK_GARAGE_DOOR_FINGERPRINTS) do
+    if
+    device:id_match(fingerprint.manufacturerId, fingerprint.productType, fingerprint.productId)
+    then
+      return true
+    end
+  end
+  return false
+end
+
+local function component_to_endpoint(device, component_id)
+  if (CONTACTSENSOR_ENDPOINT_NAME == component_id)  then
+    --contactSensor is 2 
+    return CONTACTSENSOR_ENDPOINT_NUMBER
+  end
+  -- main endpoint is garage door
+  return GDO_ENDPOINT_NUMBER
+end
+
+local function endpoint_to_component(device, ep)
+  if ( CONTACTSENSOR_ENDPOINT_NUMBER == ep ) then
+    return CONTACTSENSOR_ENDPOINT_NAME
+  end
+  return GDO_ENDPOINT_NAME
+end
+
+--- Handle BarrierOperator Report
+---
+--- @param driver st.zwave.Driver
+--- @param device st.zwave.Device
+--- @param cmd st.zwave.CommandClass.BarrierOperator.Report
+local function barrieroperator_report_handler(driver, device, cmd)
+  local barrier_event = nil
+  log.info_with
+  (
+    {hub_logs=true}, 
+    "barrieroperator_report_handler> device:%s, cmd:%s", tostring(device), tostring(cmd)
+  )
+  if cmd.args.state == BarrierOperator.state.CLOSED then
+    barrier_event = capabilities.doorControl.door.closed()
+  elseif cmd.args.state == BarrierOperator.state.CLOSING then
+    barrier_event = capabilities.doorControl.door.closing()
+  elseif cmd.args.state == BarrierOperator.state.STOPPED then
+    barrier_event = capabilities.doorControl.door.unknown()
+  elseif cmd.args.state == BarrierOperator.state.OPENING then
+    barrier_event = capabilities.doorControl.door.opening()
+  elseif cmd.args.state == BarrierOperator.state.OPEN then
+    barrier_event = capabilities.doorControl.door.open()
+  end
+
+  if (barrier_event ~= nil) then
+    device:emit_event_for_endpoint(GDO_ENDPOINT_NUMBER, barrier_event)
+  end
```

Comment:
If these events come from the GDO endpoint, you don't need to overwrite the defaults here. Since the contact sensor capability is on a different component, that event shouldn't be emitted.

Reply:
Good to know, thought this was needed. Removed.